### DISCUSSION TITLE
Support for MongoDB/MariaDB/other databases

### DISCUSSION THREAD

**Discussion by MJEND7** [ory/kratos#3007] [Q&A] [answered]
Pretty self explanatory, are you guys planning or thinking about adding support for `MongoDB` and/or any other databases besides `PostgreSQL`?

**Comment by vinckr** [comment, marked as answer]
Hello @MJEND7 

Unfortunately this is not planned at the moment.
All Ory projects support storing data in memory and in relational databases such as PostgreSQL, MySQL, SQLite and CockroachDB.

See also this discussion regarding custom database adapters
https://github.com/ory/hydra/discussions/2269
and MariaDB:
https://archive.ory.sh/t/8074603/hi-we-use-hydra-and-kratos-as-part-of-our-stackspin-project-
>MariaDB support is not planned and wont be added in the foreseeable future

**Reply by MJEND7** [reply]
Oh ok, That's a shame, thanks for the response, is there any specific documentation about creating db drivers?

**Reply by vinckr** [reply]
No, because writing your own database adapter is a very bad idea and serious security issue.

[Quoting from this comment](https://github.com/ory/hydra/issues/1627#issuecomment-549563206):

- Our internal interfaces change often, at least with every schema change. Writing your own adapter means in 99.9% of the cases that you will stay with the version your adapter works for. This is, for obvious reasons, terrible.
- The SQL datastore supports advanced logic such as ON CASCADE statements that are necessary to implement certain security-related workflows such as properly removing a chain of tokens when one of them is being revoked or a client is deleted or consent is withdrawn. We obviously have some tests for that in place internally, but not all of them are exposed and they also change depending on internal schedules.

**Reply by MJEND7** [reply]
I see, that makes sense, that is just something you can't do in a no relation database. Thanks for the in-depth response 👌

**Reply by ErgjanJaha** [reply]
There's a lot of awesome stuff you can do with mongodb aggregations!!
I'm pretty sure it would be possible, one thing thats awesome with mongodb that leaves the databases you mentioned to dust, is its ability to run stable in Kubernetes environments as a cluster, and never fail
Pg, mysql, and the rest are just really bad at that stuff! No matter which operator you use
